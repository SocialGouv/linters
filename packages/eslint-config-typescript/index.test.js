const deindent = require("deindent");
const { CLIEngine } = require("eslint");

test.each([
  [
    "@typescript-eslint/no-type-alias off",
    deindent`//
      interface Foo {
        foo: string;
      }
      interface Bar {
        bar: string;
      }
      type Qux = Foo & Bar;
    `,
  ],
  [
    "@typescript-eslint/prefer-readonly-parameter-types off",
    deindent`//
      export const array1 = (arg: string[]): sting[] => arg;
    `,
  ],
  [
    // NOTE(douglasduteil): kind of fake test
    // @typescript-eslint/typedef rule is off but we use --strict in the local tsconfig
    // see https://github.com/typescript-eslint/typescript-eslint/blob/v2.34.0/packages/eslint-plugin/docs/rules/typedef.md#requires-type-annotations-to-exist-typedef
    "@typescript-eslint/typedef off",
    deindent`//
      export const [a]: [string] = ["a"];
    `,
  ],
  [
    "import/order alphabetize",
    deindent`//
      /* eslint "@typescript-eslint/no-unused-vars": "off" */
      /* eslint "@typescript-eslint/no-unused-vars-experimental": "off" */
      /* eslint "import/no-unresolved": "off" */
      import fs from "fs";
      import path from "path";
      import chalk from "chalk";
      import _ from "lodash";
      import oof from "src/foo";
      import qux from "../../foo/qux";
      import foo from "../foo";
      import bar from "./bar";
      import baz from "./bar/baz";
      import main from "./";
    `,
  ],
])("should validate : %s", (_, code) => {
  const cli = new CLIEngine({
    useEslintrc: false,
    configFile: "index.js",
  });

  expect(cli.executeOnText(code, "index.test.js")).toMatchSnapshot();
});

test.each([
  [
    "@typescript-eslint/{no-require-import,no-var-requires} off",
    deindent`//
      const { defaults } = require("jest-config");
      const ignorePatterns = ["<rootDir>/coverage"];
      module.exports = {
        moduleDirectories: ["src", ...defaults.moduleDirectories],
        watchPathIgnorePatterns: ignorePatterns,
      };
    `,
  ],
])("jest.config.js should validate : %s", (_, code) => {
  const cli = new CLIEngine({
    useEslintrc: false,
    configFile: "index.js",
  });

  expect(cli.executeOnText(code, "jest.config.js")).toMatchSnapshot();
});

test.each([
  [
    "prettier case",
    deindent`//
    export const foo = 1
    `,
  ],
  [
    "@typescript-eslint/{no-require-imports,no-var-requires} case",
    deindent`//
    /* eslint "@typescript-eslint/no-unsafe-assignment": "off" */
    /* eslint "@typescript-eslint/no-unsafe-call": "off" */
    export var eslint = require("eslint");
    `,
  ],
  [
    "@typescript-eslint case",
    deindent`//
    type Foo = 'a' | 'b';
    `,
  ],
])("should invalidate : %s", (_, code) => {
  const cli = new CLIEngine({
    useEslintrc: false,
    configFile: "index.js",
  });

  expect(cli.executeOnText(code, "index.test.js")).toMatchSnapshot();
});
