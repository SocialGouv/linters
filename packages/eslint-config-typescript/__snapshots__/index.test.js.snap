// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jest.config.js should validate : @typescript-eslint/{no-require-import,no-var-requires} off 1`] = `
Object {
  "errorCount": 0,
  "fixableErrorCount": 0,
  "fixableWarningCount": 0,
  "results": Array [
    Object {
      "errorCount": 0,
      "filePath": Any<String>,
      "fixableErrorCount": 0,
      "fixableWarningCount": 0,
      "messages": Array [],
      "warningCount": 0,
    },
  ],
  "usedDeprecatedRules": Array [],
  "warningCount": 0,
}
`;

exports[`should invalidate : @typescript-eslint case 1`] = `
Object {
  "errorCount": 1,
  "fixableErrorCount": 1,
  "fixableWarningCount": 0,
  "results": Array [
    Object {
      "errorCount": 1,
      "filePath": Any<String>,
      "fixableErrorCount": 1,
      "fixableWarningCount": 0,
      "messages": Array [
        Object {
          "column": 12,
          "endColumn": 21,
          "endLine": 2,
          "fix": Object {
            "range": Array [
              14,
              23,
            ],
            "text": "\\"a\\" | \\"b\\"",
          },
          "line": 2,
          "message": "Replace \`'a'路|路'b'\` with \`\\"a\\"路|路\\"b\\"\`",
          "nodeType": null,
          "ruleId": "prettier/prettier",
          "severity": 2,
        },
      ],
      "source": "//
type Foo = 'a' | 'b';
",
      "warningCount": 0,
    },
  ],
  "usedDeprecatedRules": Array [],
  "warningCount": 0,
}
`;

exports[`should invalidate : @typescript-eslint/{no-require-imports,no-var-requires} case 1`] = `
Object {
  "errorCount": 2,
  "fixableErrorCount": 0,
  "fixableWarningCount": 0,
  "results": Array [
    Object {
      "errorCount": 2,
      "filePath": Any<String>,
      "fixableErrorCount": 0,
      "fixableWarningCount": 0,
      "messages": Array [
        Object {
          "column": 21,
          "endColumn": 38,
          "endLine": 4,
          "line": 4,
          "message": "Require statement not part of import statement.",
          "messageId": "noVarReqs",
          "nodeType": "CallExpression",
          "ruleId": "@typescript-eslint/no-var-requires",
          "severity": 2,
        },
        Object {
          "column": 21,
          "endColumn": 38,
          "endLine": 4,
          "line": 4,
          "message": "A \`require()\` style import is forbidden.",
          "messageId": "noRequireImports",
          "nodeType": "CallExpression",
          "ruleId": "@typescript-eslint/no-require-imports",
          "severity": 2,
        },
      ],
      "source": "//
/* eslint \\"@typescript-eslint/no-unsafe-assignment\\": \\"off\\" */
/* eslint \\"@typescript-eslint/no-unsafe-call\\": \\"off\\" */
export var eslint = require(\\"eslint\\");
",
      "warningCount": 0,
    },
  ],
  "usedDeprecatedRules": Array [],
  "warningCount": 0,
}
`;

exports[`should invalidate : prettier case 1`] = `
Object {
  "errorCount": 1,
  "fixableErrorCount": 1,
  "fixableWarningCount": 0,
  "results": Array [
    Object {
      "errorCount": 1,
      "filePath": Any<String>,
      "fixableErrorCount": 1,
      "fixableWarningCount": 0,
      "messages": Array [
        Object {
          "column": 21,
          "endColumn": 21,
          "endLine": 2,
          "fix": Object {
            "range": Array [
              23,
              23,
            ],
            "text": ";",
          },
          "line": 2,
          "message": "Insert \`;\`",
          "nodeType": null,
          "ruleId": "prettier/prettier",
          "severity": 2,
        },
      ],
      "source": "//
export const foo = 1
",
      "warningCount": 0,
    },
  ],
  "usedDeprecatedRules": Array [],
  "warningCount": 0,
}
`;

exports[`should validate : @typescript-eslint/no-type-alias off 1`] = `
Object {
  "errorCount": 0,
  "fixableErrorCount": 0,
  "fixableWarningCount": 0,
  "results": Array [
    Object {
      "errorCount": 0,
      "filePath": Any<String>,
      "fixableErrorCount": 0,
      "fixableWarningCount": 0,
      "messages": Array [],
      "warningCount": 0,
    },
  ],
  "usedDeprecatedRules": Array [],
  "warningCount": 0,
}
`;

exports[`should validate : @typescript-eslint/prefer-readonly-parameter-types off 1`] = `
Object {
  "errorCount": 0,
  "fixableErrorCount": 0,
  "fixableWarningCount": 0,
  "results": Array [
    Object {
      "errorCount": 0,
      "filePath": Any<String>,
      "fixableErrorCount": 0,
      "fixableWarningCount": 0,
      "messages": Array [],
      "warningCount": 0,
    },
  ],
  "usedDeprecatedRules": Array [],
  "warningCount": 0,
}
`;

exports[`should validate : @typescript-eslint/typedef off 1`] = `
Object {
  "errorCount": 0,
  "fixableErrorCount": 0,
  "fixableWarningCount": 0,
  "results": Array [
    Object {
      "errorCount": 0,
      "filePath": Any<String>,
      "fixableErrorCount": 0,
      "fixableWarningCount": 0,
      "messages": Array [],
      "warningCount": 0,
    },
  ],
  "usedDeprecatedRules": Array [],
  "warningCount": 0,
}
`;

exports[`should validate : import/order alphabetize 1`] = `
Object {
  "errorCount": 0,
  "fixableErrorCount": 0,
  "fixableWarningCount": 0,
  "results": Array [
    Object {
      "errorCount": 0,
      "filePath": Any<String>,
      "fixableErrorCount": 0,
      "fixableWarningCount": 0,
      "messages": Array [],
      "warningCount": 0,
    },
  ],
  "usedDeprecatedRules": Array [],
  "warningCount": 0,
}
`;
